cl/recap/mergers.py

```diff
@@ -607,28 +604,29 @@ def get_or_make_docket_entry(
      - None is returned when things fail.
     """
     if docket_entry["document_number"]:
-        with transaction.atomic():
-            # In order to create a lock, we need to retrieve the docket using
-            # select_for_update
-            docket = Docket.objects.select_for_update().get(pk=d.pk)
-            try:
-                de = DocketEntry.objects.get(
-                    docket=docket, entry_number=docket_entry["document_number"]
-                )
-                de_created = False
-            except DocketEntry.DoesNotExist:
-                de = DocketEntry.objects.create(
-                    docket=docket, entry_number=docket_entry["document_number"]
-                )
-                de_created = True
-            except DocketEntry.MultipleObjectsReturned:
-                logger.error(
-                    "Multiple docket entries found for document "
-                    "entry number '%s' while processing '%s'",
-                    docket_entry["document_number"],
-                    d,
+        try:
+            de, de_created = (
+                await DocketEntry.objects.select_related("docket")
```

Comment:
I tested this approach locally, but unfortunately is not working properly:

```
cl-django          |   File "/usr/local/lib/python3.11/site-packages/django/db/models/sql/compiler.py", line 816, in as_sql
cl-django          |     raise TransactionManagementError(
cl-django          | django.db.transaction.TransactionManagementError: select_for_update cannot be used outside of a transaction.
```

The problem is that `select_for_update` must be used within a transaction.

The tests are working because they use `TestCase`, which wraps the tests within a transaction.

> Although select_for_update() normally fails in autocommit mode, since [TestCase](https://docs.djangoproject.com/en/4.2/topics/testing/tools/#django.test.TestCase) automatically wraps each test in a transaction, calling select_for_update() in a TestCase even outside an [atomic()](https://docs.djangoproject.com/en/4.2/topics/db/transactions/#django.db.transaction.atomic) block will (perhaps unexpectedly) pass without raising a TransactionManagementError.

Since a transaction can't be used in an async context I think a workaround that can help is to perform the transaction synchronous.

```
def add_create_docket_entry_transaction(d, docket_entry):
    with transaction.atomic():
        de, de_created = (
            DocketEntry.objects.select_related("docket")
            .select_for_update(of=("self", "docket"))
            .get_or_create(
                docket=d, entry_number=docket_entry["document_number"]
            )
        )
        return de, de_created
```

And then call it on `get_or_make_docket_entry` using `sync_to_async:`
`de, de_created = await sync_to_async(add_create_docket_entry_transaction)(d, docket_entry)`

Reply:
> The problem is that `select_for_update` must be used within a transaction.

Ugh, that's annoying.

> Since a transaction can't be used in an async context I think a workaround that can help is to perform the transaction synchronous.

Ok, I've gone ahead and made that along with the duplicate error handling code path synchronous along the lines of your suggestion.

Reply:
Thanks. I think this is pretty close. I tested this new approach against the race condition, but it's still occurring. So, I've made some suggestions in the comments below.